File: lib-sdk/flutter-sdk/lib/src/tracker.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;
import 'package:uuid/uuid.dart';
import 'models/models.dart';

abstract class DirectoAiTracker {
  Future<void> trackEvent(String name, Map<String, dynamic> data);
  Future<void> trackImpression(String contentId, Map<String, dynamic> data);
  Future<void> trackViewTime(
    String contentId,
    int seconds,
    Map<String, dynamic> data,
  );
  Future<void> toggleLike(String contentId, {String? campaignId});
  Future<List<Post>> fetchProfileFeed(String profileAccountId);
  Future<void> followAccount(String profileAccountId);
  Future<void> unfollowAccount(String profileAccountId);
  Future<void> toggleFollowAccount(String profileAccountId, bool isFollowing);
  Future<void> addFavorite(String contentId, {String? campaignId});
  Future<void> removeFavorite(String contentId);
  Future<void> toggleFavorite(
    String contentId,
    bool isFavorited, {
    String? campaignId,
  });
  Future<void> shareContent(
    Map<String, dynamic> contentData, {
    String? campaignId,
    String? title,
  });
  Future<void> reportContent(
    String contentId, {
    required String reportType,
    String? description,
  });
}

class NoopDirectoAiTracker implements DirectoAiTracker {
  @override
  Future<void> addFavorite(String contentId, {String? campaignId}) async {}

  @override
  Future<List<Post>> fetchProfileFeed(String profileAccountId) async {
    return <Post>[];
  }

  @override
  Future<void> followAccount(String profileAccountId) async {}

  @override
  Future<void> removeFavorite(String contentId) async {}

  @override
  Future<void> reportContent(
    String contentId, {
    required String reportType,
    String? description,
  }) async {}

  @override
  Future<void> shareContent(
    Map<String, dynamic> contentData, {
    String? campaignId,
    String? title,
  }) async {}

  @override
  Future<void> toggleFavorite(
    String contentId,
    bool isFavorited, {
    String? campaignId,
  }) async {}

  @override
  Future<void> toggleFollowAccount(
    String profileAccountId,
    bool isFollowing,
  ) async {}

  @override
  Future<void> toggleLike(String contentId, {String? campaignId}) async {}

  @override
  Future<void> trackEvent(String name, Map<String, dynamic> data) async {}

  @override
  Future<void> trackImpression(
    String contentId,
    Map<String, dynamic> data,
  ) async {}

  @override
  Future<void> trackViewTime(
    String contentId,
    int seconds,
    Map<String, dynamic> data,
  ) async {}

  @override
  Future<void> unfollowAccount(String profileAccountId) async {}
}

class DefaultDirectoAiTracker implements DirectoAiTracker {
  final DirectoAiConfig config;
  final _uuid = const Uuid();

  DefaultDirectoAiTracker(this.config);

  String get baseUrl => config.baseUrl ?? "https://api.directoai.com.br";

  Uint8List _stringToUTF16LE(String value) {
    final bytes = Uint8List(value.length * 2);
    for (var i = 0; i < value.length; i++) {
      final code = value.codeUnitAt(i);
      bytes[i * 2] = code & 0xFF;
      bytes[i * 2 + 1] = code >> 8;
    }
    return bytes;
  }

  Future<String> _computeSHA256(String value) async {
    if (value.isEmpty) return '';

    final utf16Bytes = _stringToUTF16LE(value);
    return sha256.convert(utf16Bytes).toString().toUpperCase();
  }

  Future<void> _sendToAnalytics(
    String eventName,
    Map<String, dynamic> eventParams,
  ) async {
    if (config.onAnalyticsEvent != null) {
      try {
        config.onAnalyticsEvent!(eventName, eventParams);
      } catch (e) {
        print('DirectoAi SDK: Error in analytics callback: $e');
      }
    }
  }

  Future<void> _sendMessageQueue(Map<String, dynamic> payload) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/metric-queue'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(payload),
      );
      if (response.statusCode >= 400) {
        print(
          'DirectoAi SDK: Error sending message to queue: ${response.statusCode}',
        );
      }
    } catch (e) {
      print('DirectoAi SDK: Failed to send analytic event: $e');
    }
  }

  @override
  Future<void> trackEvent(String name, Map<String, dynamic> data) async {
    final payload = {
      'type': 'click',
      'data': {
        ...data,
        'accountId': config.accountId,
        'customerId': config.customerId,
        'deviceId': config.deviceId,
        'eventType': name,
        'createdAt': DateTime.now().toUtc().toIso8601String(),
      },
    };
    return _sendMessageQueue(payload);
  }

  @override
  Future<void> trackImpression(
    String contentId,
    Map<String, dynamic> data,
  ) async {
    final payload = {
      'type': 'view',
      'data': {
        ...data,
        'contentId': contentId,
        'accountId': config.accountId,
        'customerId': config.customerId,
        'deviceId': config.deviceId,
        'createdAt': DateTime.now().toUtc().toIso8601String(),
      },
    };
    return _sendMessageQueue(payload);
  }

  @override
  Future<void> trackViewTime(
    String contentId,
    int seconds,
    Map<String, dynamic> data,
  ) async {
    final payload = {
      'type': 'timeView',
      'data': {
        ...data,
        'contentId': contentId,
        'time': seconds,
        'accountId': config.accountId,
        'customerId': config.customerId,
        'deviceId': config.deviceId,
        'createdAt': DateTime.now().toUtc().toIso8601String(),
      },
    };
    return _sendMessageQueue(payload);
  }

  @override
  Future<void> toggleLike(String contentId, {String? campaignId}) async {
    return trackEvent('click-like', {
      'contentId': contentId,
      if (campaignId != null) 'campaignId': campaignId,
    });
  }

  @override
  Future<List<Post>> fetchProfileFeed(String profileAccountId) async {
    if (config.accountId.isEmpty || config.apiKey.isEmpty) {
      throw Exception('Missing required data for profile feed request');
    }

    final url =
        '$baseUrl/campaign/api/v1/feed/accounts?accountId=${Uri.encodeQueryComponent(config.accountId)}&apiKey=${Uri.encodeQueryComponent(config.apiKey)}&profileAccountId=${Uri.encodeQueryComponent(profileAccountId)}';

    try {
      final response = await http.get(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode >= 300) {
        throw Exception(
          'Profile feed request failed with status: ${response.statusCode}',
        );
      }

      final payload = jsonDecode(response.body);
      final dynamic data = payload['data'];
      final dynamic feeds = data is Map<String, dynamic>
          ? data['feeds']
          : payload['feeds'] ?? data;

      if (feeds is! List) {
        return <Post>[];
      }

      return feeds
          .whereType<Map>()
          .map((item) => Post.fromJson(Map<String, dynamic>.from(item)))
          .toList();
    } catch (e) {
      print('DirectoAi SDK: Failed to fetch profile feed: $e');
      rethrow;
    }
  }

  @override
  Future<void> followAccount(String profileAccountId) async {
    if (config.accountId.isEmpty ||
        config.apiKey.isEmpty ||
        (config.customerId?.isEmpty ?? true)) {
      throw Exception('Missing required data for follow action');
    }

    final url =
        '$baseUrl/campaign/api/v1/feed/followers?accountId=${Uri.encodeQueryComponent(config.accountId)}&apiKey=${Uri.encodeQueryComponent(config.apiKey)}';

    try {
      final response = await http.post(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'accountId': profileAccountId,
          'customerId': config.customerId,
        }),
      );

      if (response.statusCode >= 300) {
        throw Exception(
          'Follow account failed with status: ${response.statusCode}',
        );
      }

      await trackEvent('click-follow', {'profileAccountId': profileAccountId});
    } catch (e) {
      print('DirectoAi SDK: Failed to follow account: $e');
      rethrow;
    }
  }

  @override
  Future<void> unfollowAccount(String profileAccountId) async {
    if (config.accountId.isEmpty ||
        config.apiKey.isEmpty ||
        (config.customerId?.isEmpty ?? true)) {
      throw Exception('Missing required data for unfollow action');
    }

    final url =
        '$baseUrl/campaign/api/v1/feed/followers?accountId=${Uri.encodeQueryComponent(config.accountId)}&apiKey=${Uri.encodeQueryComponent(config.apiKey)}';

    try {
      final response = await http.delete(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'accountId': profileAccountId,
          'customerId': config.customerId,
        }),
      );

      if (response.statusCode >= 300) {
        throw Exception(
          'Unfollow account failed with status: ${response.statusCode}',
        );
      }

      await trackEvent('click-unfollow', {
        'profileAccountId': profileAccountId,
      });
    } catch (e) {
      print('DirectoAi SDK: Failed to unfollow account: $e');
      rethrow;
    }
  }

  @override
  Future<void> toggleFollowAccount(
    String profileAccountId,
    bool isFollowing,
  ) async {
    if (isFollowing) {
      return unfollowAccount(profileAccountId);
    }

    return followAccount(profileAccountId);
  }

  @override
  Future<void> addFavorite(String contentId, {String? campaignId}) async {
    final url = '$baseUrl/campaign/api/v1/feed/favorites';

    try {
      final response = await http.post(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'accountId': config.accountId,
          'customerId': config.customerId,
          'contentId': contentId,
          if (campaignId != null) 'campaignId': campaignId,
        }),
      );

      // Log metrics on success or conflict (already favorited)
      if (response.statusCode < 300 || response.statusCode == 409) {
        await trackEvent('click-favorite', {
          'contentId': contentId,
          if (campaignId != null) 'campaignId': campaignId,
        });

        await _sendToAnalytics('click_card', {
          'content_id': contentId,
          'account_id': config.accountId,
          'customer_id': config.customerId,
          'event_type': 'click-favorite',
        });
      }

      if (response.statusCode >= 300 && response.statusCode != 409) {
        throw Exception(
          'Add favorite failed with status: ${response.statusCode}',
        );
      }
    } catch (e) {
      print('DirectoAi SDK: Error adding favorite: $e');
      rethrow;
    }
  }

  @override
  Future<void> removeFavorite(String contentId) async {
    final url = '$baseUrl/campaign/api/v1/feed/favorites';

    try {
      final response = await http.delete(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'accountId': config.accountId,
          'customerId': config.customerId,
          'contentId': contentId,
        }),
      );

      if (response.statusCode >= 300) {
        throw Exception(
          'Remove favorite failed with status: ${response.statusCode}',
        );
      }
    } catch (e) {
      print('DirectoAi SDK: Error removing favorite: $e');
      rethrow;
    }
  }

  @override
  Future<void> toggleFavorite(
    String contentId,
    bool isFavorited, {
    String? campaignId,
  }) async {
    if (isFavorited) {
      return removeFavorite(contentId);
    } else {
      return addFavorite(contentId, campaignId: campaignId);
    }
  }

  @override
  Future<void> shareContent(
    Map<String, dynamic> contentData, {
    String? campaignId,
    String? title,
  }) async {
    final shareId = _uuid.v4();
    final contentId =
        contentData['contentId']?.toString() ??
        contentData['id']?.toString() ??
        '';

    String? base64Template;
    if (contentData['template'] != null && contentData['template'] is String) {
      try {
        final bytes = utf8.encode(contentData['template']);
        base64Template = base64Encode(bytes);
      } catch (_) {}
    }

    try {
      final payload = {
        'shareId': shareId,
        'contentId': contentId,
        'campaignId': campaignId ?? '',
        'accountId': config.accountId,
        'createdBy': config.customerId ?? '',
        'expiresInHours': 168,
        'content': {
          ...contentData,
          'encryptedSnapshot': base64Template,
          'template': null,
        },
      };

      print('DirectoAi SDK: Creating share link...');
      print('URL: $baseUrl/campaign/api/v1/feed/share');
      // print('Payload: ${jsonEncode(payload)}'); // Omit for brevity in logs unless needed

      final response = await http.post(
        Uri.parse('$baseUrl/campaign/api/v1/feed/share'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(payload),
      );

      if (response.statusCode < 300) {
        final result = jsonDecode(response.body);
        final publicShareUrl = result['data']?['url']?.toString();

        if (publicShareUrl == null) {
          throw Exception('Share API failed: No URL returned');
        }

        print('DirectoAi SDK: Share created successfully: $publicShareUrl');

        await trackEvent('click-share', {
          'contentId': contentId,
          if (campaignId != null) 'campaignId': campaignId,
        });
      } else {
        throw Exception(
          'Share API failed with status: ${response.statusCode} - ${response.body}',
        );
      }
    } catch (e) {
      print('DirectoAi SDK: Error creating share link: $e');
    }
  }

  @override
  Future<void> reportContent(
    String contentId, {
    required String reportType,
    String? description,
  }) async {
    final url = '$baseUrl/content/api/v1/contents/$contentId/report';
    final reporterCustomerId = await _computeSHA256(config.customerId ?? '');

    try {
      final response = await http.post(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'reportType': reportType,
          'reporterCustomerId': reporterCustomerId,
          if (description != null && description.isNotEmpty)
            'description': description,
        }),
      );

      if (response.statusCode >= 300) {
        throw Exception(
          'Report content failed with status: ${response.statusCode}',
        );
      }

      await trackEvent('click-report', {
        'contentId': contentId,
        'reportType': reportType,
      });
    } catch (e) {
      print('DirectoAi SDK: Error reporting content: $e');
      rethrow;
    }
  }
}

File: lib-sdk/react-lib/src/core/tracker.ts
import type { DirectoAiConfig, Post } from "./types.js";

export interface DirectoAiTracker {
  trackEvent(name: string, data: Record<string, any>): Promise<void>;
  trackImpression(contentId: string, data: Record<string, any>): Promise<void>;
  trackViewTime(
    contentId: string,
    seconds: number,
    data: Record<string, any>,
  ): Promise<void>;
  toggleLike(contentId: string, campaignId?: string): Promise<void>;
  fetchProfileFeed(profileAccountId: string): Promise<Post[]>;
  followAccount(profileAccountId: string): Promise<void>;
  unfollowAccount(profileAccountId: string): Promise<void>;
  toggleFollowAccount(
    profileAccountId: string,
    isFollowing: boolean,
  ): Promise<void>;
  addFavorite(contentId: string, campaignId?: string): Promise<void>;
  removeFavorite(contentId: string): Promise<void>;
  toggleFavorite(
    contentId: string,
    campaignId: string | undefined,
    isFavorited: boolean,
  ): Promise<void>;
  reportContent(
    contentId: string,
    reportType: string,
    description?: string,
  ): Promise<void>;
  shareContent(contentData: any): Promise<void>;
}

export class NoopDirectoAiTracker implements DirectoAiTracker {
  async trackEvent(_name: string, _data: Record<string, any>): Promise<void> {}

  async trackImpression(
    _contentId: string,
    _data: Record<string, any>,
  ): Promise<void> {}

  async trackViewTime(
    _contentId: string,
    _seconds: number,
    _data: Record<string, any>,
  ): Promise<void> {}

  async toggleLike(
    _contentId: string,
    _campaignId?: string,
  ): Promise<void> {}

  async fetchProfileFeed(_profileAccountId: string): Promise<Post[]> {
    return [];
  }

  async followAccount(_profileAccountId: string): Promise<void> {}

  async unfollowAccount(_profileAccountId: string): Promise<void> {}

  async toggleFollowAccount(
    _profileAccountId: string,
    _isFollowing: boolean,
  ): Promise<void> {}

  async addFavorite(
    _contentId: string,
    _campaignId?: string,
  ): Promise<void> {}

  async removeFavorite(_contentId: string): Promise<void> {}

  async toggleFavorite(
    _contentId: string,
    _campaignId: string | undefined,
    _isFavorited: boolean,
  ): Promise<void> {}

  async reportContent(
    _contentId: string,
    _reportType: string,
    _description?: string,
  ): Promise<void> {}

  async shareContent(_contentData: any): Promise<void> {}
}

export class DefaultDirectoAiTracker implements DirectoAiTracker {
  private config: DirectoAiConfig;

  constructor(config: DirectoAiConfig) {
    this.config = config;
  }

  private get baseUrl() {
    return this.config.baseUrl || "https://api.directoai.com.br";
  }

  private stringToUTF16LE(value: string) {
    const bytes = new Uint8Array(value.length * 2);
    for (let i = 0; i < value.length; i += 1) {
      const code = value.charCodeAt(i);
      bytes[i * 2] = code & 0xff;
      bytes[i * 2 + 1] = code >> 8;
    }
    return bytes;
  }

  private async computeSHA256(value: string) {
    if (!value) return "";

    const utf16Bytes = this.stringToUTF16LE(value);
    const hashBuffer = await crypto.subtle.digest("SHA-256", utf16Bytes);
    const hashArray = new Uint8Array(hashBuffer);

    return Array.from(hashArray)
      .map((byte) => byte.toString(16).padStart(2, "0"))
      .join("")
      .toUpperCase();
  }

  private async sendToGA(eventName: string, eventParams: Record<string, any>) {
    if (this.config.googleAnalyticsHandler) {
      try {
        this.config.googleAnalyticsHandler(eventName, eventParams);
      } catch (e) {
        console.error("DirectoAi SDK: Error in GA handler", e);
      }
    }
  }

  private async sendMessageQueue(payload: object) {
    try {
      const response = await fetch(`${this.baseUrl}/metric-queue`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        console.warn(
          "DirectoAi SDK: Error sending message to queue",
          response.status,
        );
      }
    } catch (error) {
      console.error("DirectoAi SDK: Failed to send analytic event:", error);
    }
  }

  async trackEvent(name: string, data: Record<string, any>) {
    const payload = {
      type: "click",
      data: {
        ...data,
        accountId: this.config.accountId,
        customerId: this.config.customerId,
        deviceId: this.config.deviceId,
        eventType: name,
        createdAt: new Date().toISOString(),
      },
    };
    return this.sendMessageQueue(payload);
  }

  async trackImpression(contentId: string, data: Record<string, any>) {
    const payload = {
      type: "view",
      data: {
        ...data,
        contentId,
        accountId: this.config.accountId,
        customerId: this.config.customerId,
        deviceId: this.config.deviceId,
        createdAt: new Date().toISOString(),
      },
    };
    return this.sendMessageQueue(payload);
  }

  async trackViewTime(
    contentId: string,
    seconds: number,
    data: Record<string, any>,
  ) {
    const payload = {
      type: "timeView",
      data: {
        ...data,
        contentId,
        time: seconds,
        accountId: this.config.accountId,
        customerId: this.config.customerId,
        deviceId: this.config.deviceId,
        createdAt: new Date().toISOString(),
      },
    };
    return this.sendMessageQueue(payload);
  }

  async toggleLike(contentId: string, campaignId?: string) {
    return this.trackEvent("click-like", {
      contentId,
      campaignId,
    });
  }

  async fetchProfileFeed(profileAccountId: string): Promise<Post[]> {
    if (!this.config.accountId || !this.config.apiKey) {
      throw new Error("Missing required data for profile feed request");
    }

    const url = `${this.baseUrl}/campaign/api/v1/feed/accounts?accountId=${encodeURIComponent(this.config.accountId)}&apiKey=${encodeURIComponent(this.config.apiKey)}&profileAccountId=${encodeURIComponent(profileAccountId)}`;

    try {
      const response = await fetch(url, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        throw new Error(`Profile feed request failed: ${response.status}`);
      }

      const payload = await response.json();
      const feeds = Array.isArray(payload?.data?.feeds)
        ? payload.data.feeds
        : Array.isArray(payload?.data)
          ? payload.data
          : Array.isArray(payload?.feeds)
            ? payload.feeds
            : [];

      return feeds as Post[];
    } catch (error) {
      console.error("DirectoAi SDK: Failed to fetch profile feed:", error);
      throw error;
    }
  }

  async followAccount(profileAccountId: string) {
    if (!this.config.accountId || !this.config.apiKey || !this.config.customerId) {
      throw new Error("Missing required data for follow action");
    }

    const url = `${this.baseUrl}/campaign/api/v1/feed/followers?accountId=${encodeURIComponent(this.config.accountId)}&apiKey=${encodeURIComponent(this.config.apiKey)}`;

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          accountId: profileAccountId,
          customerId: this.config.customerId,
        }),
      });

      if (!response.ok) {
        throw new Error(`Follow account failed: ${response.status}`);
      }

      await this.trackEvent("click-follow", {
        profileAccountId,
      });
    } catch (error) {
      console.error("DirectoAi SDK: Failed to follow account:", error);
      throw error;
    }
  }

  async unfollowAccount(profileAccountId: string) {
    if (!this.config.accountId || !this.config.apiKey || !this.config.customerId) {
      throw new Error("Missing required data for unfollow action");
    }

    const url = `${this.baseUrl}/campaign/api/v1/feed/followers?accountId=${encodeURIComponent(this.config.accountId)}&apiKey=${encodeURIComponent(this.config.apiKey)}`;

    try {
      const response = await fetch(url, {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          accountId: profileAccountId,
          customerId: this.config.customerId,
        }),
      });

      if (!response.ok) {
        throw new Error(`Unfollow account failed: ${response.status}`);
      }

      await this.trackEvent("click-unfollow", {
        profileAccountId,
      });
    } catch (error) {
      console.error("DirectoAi SDK: Failed to unfollow account:", error);
      throw error;
    }
  }

  async toggleFollowAccount(profileAccountId: string, isFollowing: boolean) {
    if (isFollowing) {
      return this.unfollowAccount(profileAccountId);
    }

    return this.followAccount(profileAccountId);
  }

  async addFavorite(contentId: string, campaignId?: string) {
    const url = `${this.baseUrl}/campaign/api/v1/feed/favorites`;
    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          accountId: this.config.accountId,
          customerId: this.config.customerId,
          contentId,
          campaignId,
        }),
      });

      // Se sucesso ou conflito (já favoritado), logamos as métricas
      if (response.ok || response.status === 409) {
        await this.trackEvent("click-favorite", {
          contentId,
          campaignId,
        });

        await this.sendToGA("click_card", {
          content_id: contentId,
          account_id: this.config.accountId,
          customer_id: this.config.customerId,
          event_type: "click-favorite",
        });
      }

      if (!response.ok && response.status !== 409) {
        throw new Error(`Add favorite failed: ${response.status}`);
      }
    } catch (error) {
      console.error("DirectoAi SDK: Failed to add favorite:", error);
      throw error;
    }
  }

  async removeFavorite(contentId: string) {
    const url = `${this.baseUrl}/campaign/api/v1/feed/favorites`;
    try {
      const response = await fetch(url, {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          accountId: this.config.accountId,
          customerId: this.config.customerId,
          contentId,
        }),
      });

      if (!response.ok) {
        throw new Error(`Remove favorite failed: ${response.status}`);
      }
    } catch (error) {
      console.error("DirectoAi SDK: Failed to remove favorite:", error);
      throw error;
    }
  }

  async toggleFavorite(
    contentId: string,
    campaignId: string | undefined,
    isFavorited: boolean,
  ) {
    if (isFavorited) {
      return this.removeFavorite(contentId);
    } else {
      return this.addFavorite(contentId, campaignId);
    }
  }

  async reportContent(
    contentId: string,
    reportType: string,
    description?: string,
  ) {
    const url = `${this.baseUrl}/content/api/v1/contents/${contentId}/report`;
    const reporterCustomerId = await this.computeSHA256(
      this.config.customerId || "",
    );

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          reportType,
          reporterCustomerId,
          ...(description ? { description } : {}),
        }),
      });

      if (!response.ok) {
        throw new Error(`Report content failed: ${response.status}`);
      }

      await this.trackEvent("click-report", {
        contentId,
        reportType,
      });
    } catch (error) {
      console.error("DirectoAi SDK: Failed to report content:", error);
      throw error;
    }
  }

  async shareContent(contentData: any) {
    const generateUUID = () => {
      try {
        if (typeof crypto !== "undefined" && crypto.randomUUID) {
          return crypto.randomUUID();
        }
        // Fallback for non-secure contexts or older environments
        return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
          const r = (Math.random() * 16) | 0;
          const v = c === "x" ? r : (r & 0x3) | 0x8;
          return v.toString(16);
        });
      } catch (e) {
        return Date.now().toString(36) + Math.random().toString(36).substr(2);
      }
    };

    const shareId = generateUUID();
    const contentId = contentData?.contentId || contentData?.id || "";
    const campaignId = contentData?.campaignId || "";

    const base64Encode = (str: string) => {
      try {
        if (typeof window !== "undefined") {
          return window.btoa(unescape(encodeURIComponent(str)));
        } else if (typeof globalThis !== "undefined" && globalThis.btoa) {
          return globalThis.btoa(unescape(encodeURIComponent(str)));
        } else {
          return (globalThis as any).Buffer.from(str).toString("base64");
        }
      } catch (e) {
        return "";
      }
    };

    try {
      const payload = {
        shareId: shareId,
        contentId: contentId,
        campaignId: campaignId,
        accountId: this.config.accountId,
        createdBy: this.config.customerId || "",
        expiresInHours: 168,
        content: {
          ...contentData,
          encryptedSnapshot: contentData?.template
            ? base64Encode(contentData.template)
            : null,
          template: null,
        },
      };

      console.log("DirectoAi SDK: Creating share link...", {
        url: `${this.baseUrl}/campaign/api/v1/feed/share`,
        payload,
      });

      const response = await fetch(
        `${this.baseUrl}/campaign/api/v1/feed/share`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(payload),
        },
      );

      if (response.ok) {
        const result = await response.json();
        const publicShareUrl = result.data?.url;

        if (!publicShareUrl) {
          throw new Error("Share API failed: No URL returned");
        }

        console.log("DirectoAi SDK: Share created successfully:", result);

        await this.trackEvent("click-share", {
          contentId,
          campaignId,
        });

        const shareData = {
          title: contentData?.title || "Compartilhar",
          url: publicShareUrl,
          text: contentData?.title || "",
        };

        if ((window as any).flutter_inappwebview?.callHandler) {
          (window as any).flutter_inappwebview.callHandler(
            "shareLink",
            publicShareUrl,
          );
        } else if (
          navigator.share &&
          navigator.canShare &&
          navigator.canShare(shareData)
        ) {
          await navigator.share(shareData);
        } else {
          await navigator.clipboard.writeText(publicShareUrl);
        }
      } else {
        const errorText = await response.text();
        throw new Error(
          `Share API failed: ${response.status} ${response.statusText} - ${errorText}`,
        );
      }
    } catch (error) {
      console.error("DirectoAi SDK: Error creating share link:", error);
    }
  }
}

/**
 * Standalone utility to create a share link, matching the pattern provided by the user.
 */
export async function createShareLink(
  config: DirectoAiConfig,
  content: any,
  campaignId: string | null = "",
): Promise<string | null> {
  const generateUUID = () => {
    try {
      if (typeof crypto !== "undefined" && crypto.randomUUID) {
        return crypto.randomUUID();
      }
      return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
        const r = (Math.random() * 16) | 0;
        const v = c === "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
      });
    } catch (e) {
      return Date.now().toString(36) + Math.random().toString(36).substr(2);
    }
  };

  const base64Encode = (str: string) => {
    try {
      if (typeof window !== "undefined") {
        return window.btoa(unescape(encodeURIComponent(str)));
      } else if (typeof globalThis !== "undefined" && globalThis.btoa) {
        return globalThis.btoa(unescape(encodeURIComponent(str)));
      } else {
        return (globalThis as any).Buffer.from(str).toString("base64");
      }
    } catch (e) {
      return "";
    }
  };

  try {
    const shareId = generateUUID();
    const payload = {
      shareId,
      contentId: content.contentId || content.id || "",
      campaignId: campaignId || "",
      accountId: config.accountId,
      createdBy: config.customerId || "",
      expiresInHours: 168,
      content: {
        ...content,
        encryptedSnapshot: content.template
          ? base64Encode(content.template)
          : null,
        template: null,
      },
    };

    const baseUrl = config.baseUrl || "https://api.directoai.com.br";
    const response = await fetch(`${baseUrl}/campaign/api/v1/feed/share`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(
        `Share API failed: ${response.status} ${response.statusText}`,
      );
    }

    const result = await response.json();
    return result.data?.url || null;
  } catch (error) {
    console.error("Error creating share link:", error);
    return null;
  }
}
